File: lib/DiscretePID.cpp
#include "DiscretePID.h"
#include <algorithm>
#include <cmath>

namespace ctrl
{

    DiscretePID::DiscretePID(const PIDParams &params, double sampleTime)
        : p_(params), Ts_(sampleTime)
    {
        reset();
    }

    // Discrete PID - backward Euler integral + backward-Euler filtered derivative.
    //
    // Derivative filter recurrence (derived from C_D(s) = Kd.N.s/(s+N), backward Euler):
    //   alpha     = 1 / (1 + N.Ts)
    //   d[k]  = alpha.d[k-1] + Kd.N.alpha.(e[k] - e[k-1])
    //
    // Anti-windup back-calculation (Astrom & Wittenmark):
    //   I[k+1] = I[k] + Ki.Ts.e[k] + Kb.(u_sat[k] - u_unsat[k])
    double DiscretePID::compute(double error)
    {
        // Hold last output on bad measurement rather than corrupting integral state
        if (!std::isfinite(error))
            return u_prev_;

        // Filtered derivative
        const double alpha = 1.0 / (1.0 + p_.N * Ts_);
        const double d_new = alpha * deriv_ + p_.Kd * p_.N * alpha * (error - e_prev_);

        // Integral: backward Euler increment for this step
        const double ki_update = p_.Ki * Ts_ * error;

        // Unsaturated output - uses I[k] = I[k-1] + ki_update (backward Euler)
        const double u_unsat = p_.Kp * error + (integral_ + ki_update) + d_new;

        // Output saturation
        const double u_sat = std::max(p_.uMin, std::min(p_.uMax, u_unsat));

        // Integral state update with anti-windup back-calculation
        integral_ += ki_update + p_.Kb * (u_sat - u_unsat);

        // State updates
        deriv_ = d_new;
        e_prev_ = error;
        u_prev_ = u_sat;

        return u_sat;
    }

    void DiscretePID::reset()
    {
        integral_ = 0.0;
        deriv_ = 0.0;
        e_prev_ = 0.0;
        y_prev_ = 0.0;
        u_prev_ = 0.0;
    }

    // Back-calculate the integral state so that compute(error) approx = u_target.
    // Derivation: u_target approx = Kp*error + integral + Kd*N*(error-e_prev)/(1+N*Ts)
    // Solving for integral: integral = u_target - Kp*error - deriv (zeroed here)
    // On the very first compute() call after this, the output will be close to u_target.
    void DiscretePID::bumplessInit(double u_target, double error)
    {
        deriv_    = 0.0;
        e_prev_   = error;
        u_prev_   = u_target;
        integral_ = u_target - p_.Kp * error;  // absorb P term; D starts from rest
    }

    // Derivative on measurement: same as compute() but routes the derivative filter
    // through -y (plant output) instead of e = r - y. This eliminates the derivative
    // spike that occurs when the setpoint r steps.
    //
    // Derivation:
    //   Standard: d[k] = alpha*d[k-1] + Kd*N*alpha*(e[k] - e[k-1])
    //   DoM:      d[k] = alpha*d[k-1] - Kd*N*alpha*(y[k] - y[k-1])   <- derivative of -y only
    //   (sign correct: for a step up in r, y increases, d[k] goes negative = less overshoot)
    //
    // Equivalent MATLAB: Discrete PID Controller block, DerivativeFilter on 'Measurement'.
    double DiscretePID::computeDoM(double y, double r)
    {
        const double error = r - y;

        if (!std::isfinite(error) || !std::isfinite(y))
            return u_prev_;

        // Derivative filter on -y (not on error)
        const double alpha = 1.0 / (1.0 + p_.N * Ts_);
        const double d_new = alpha * deriv_ - p_.Kd * p_.N * alpha * (y - y_prev_);

        // Integral: backward Euler on error (unchanged)
        const double ki_update = p_.Ki * Ts_ * error;

        // Unsaturated output
        const double u_unsat = p_.Kp * error + (integral_ + ki_update) + d_new;

        // Output saturation
        const double u_sat = std::max(p_.uMin, std::min(p_.uMax, u_unsat));

        // Integral state update with anti-windup back-calculation
        integral_ += ki_update + p_.Kb * (u_sat - u_unsat);

        // State updates
        deriv_  = d_new;
        e_prev_ = error;
        y_prev_ = y;
        u_prev_ = u_sat;

        return u_sat;
    }

} // namespace ctrl

File: MATLAB/+ctrl/DiscretePID.m
classdef DiscretePID < handle
%DISCRETEPID  Discrete-time PID with filtered derivative and back-calculation anti-windup.
%   Faithful port of ctrl::DiscretePID (lib/DiscretePID.{h,cpp}).
%
%   Construction:
%       pid = ctrl.DiscretePID(params, Ts)
%   where params is a struct with any subset of the fields returned by
%   ctrl.DiscretePID.defaultParams():
%       Kp, Ki, Kd, N, uMin, uMax, Kb, b_weight
%   (missing fields take their defaults, matching the C++ PIDParams struct).
%
%   Sign convention (compute): tracking error e = r - y.
%   Discretisation: backward Euler for both the integral and the derivative filter.
%   NaN contract: holds the last output on a non-finite input.

  properties
    p        % struct of tuning parameters
    Ts       % sample time [s]
    integral % accumulated integral I[k]
    deriv    % filtered derivative state d[k-1]
    e_prev   % previous error e[k-1]
    y_prev   % previous measurement y[k-1] (derivative-on-measurement)
    u_prev   % previous saturated output u[k-1] (anti-windup)
  end

  methods
    function obj = DiscretePID(params, sampleTime)
      obj.p = ctrl.DiscretePID.defaultParams();
      if nargin >= 1 && ~isempty(params)
        fn = fieldnames(params);
        for i = 1:numel(fn)
          obj.p.(fn{i}) = params.(fn{i});
        end
      end
      obj.Ts = sampleTime;
      obj.reset();
    end

    function u = compute(obj, error)
      %COMPUTE  u[k] from tracking error e[k] = r[k] - y[k].
      if ~isfinite(error)
        u = obj.u_prev;   % hold last output on bad measurement
        return;
      end

      % Filtered derivative (backward Euler): alpha = 1/(1+N Ts)
      alpha = 1.0 / (1.0 + obj.p.N * obj.Ts);
      d_new = alpha * obj.deriv + obj.p.Kd * obj.p.N * alpha * (error - obj.e_prev);

      % Integral: backward-Euler increment for this step
      ki_update = obj.p.Ki * obj.Ts * error;

      % Unsaturated output uses I[k] = I[k-1] + ki_update
      u_unsat = obj.p.Kp * error + (obj.integral + ki_update) + d_new;

      % Output saturation
      u_sat = max(obj.p.uMin, min(obj.p.uMax, u_unsat));

      % Integral state update with anti-windup back-calculation
      obj.integral = obj.integral + ki_update + obj.p.Kb * (u_sat - u_unsat);

      obj.deriv  = d_new;
      obj.e_prev = error;
      obj.u_prev = u_sat;
      u = u_sat;
    end

    function u = computeDoM(obj, y, r)
      %COMPUTEDOM  Derivative-on-measurement variant (no derivative kick on setpoint steps).
      error = r - y;
      if ~isfinite(error) || ~isfinite(y)
        u = obj.u_prev;
        return;
      end

      alpha = 1.0 / (1.0 + obj.p.N * obj.Ts);
      % Derivative filter on -y (not on error)
      d_new = alpha * obj.deriv - obj.p.Kd * obj.p.N * alpha * (y - obj.y_prev);

      ki_update   = obj.p.Ki * obj.Ts * error;
      prop_signal = obj.p.b_weight * r - y;   % 2DOF setpoint weight
      u_unsat     = obj.p.Kp * prop_signal + (obj.integral + ki_update) + d_new;
      u_sat       = max(obj.p.uMin, min(obj.p.uMax, u_unsat));

      obj.integral = obj.integral + ki_update + obj.p.Kb * (u_sat - u_unsat);

      obj.deriv  = d_new;
      obj.e_prev = error;
      obj.y_prev = y;
      obj.u_prev = u_sat;
      u = u_sat;
    end

    function reset(obj)
      obj.integral = 0.0;
      obj.deriv    = 0.0;
      obj.e_prev   = 0.0;
      obj.y_prev   = 0.0;
      obj.u_prev   = 0.0;
    end

    function bumplessInit(obj, u_target, error)
      %BUMPLESSINIT  Set integral so the next compute(error) returns ~u_target.
      obj.deriv    = 0.0;
      obj.e_prev   = error;
      obj.u_prev   = u_target;
      obj.integral = u_target - obj.p.Kp * error;
    end

    function t = sampleTime(obj), t = obj.Ts; end
    function u = lastOutput(obj), u = obj.u_prev; end
    function tf = hasInternalAntiWindup(obj), tf = (obj.p.Kb ~= 0.0); end
  end

  methods (Static)
    function p = defaultParams()
      %DEFAULTPARAMS  PIDParams defaults matching lib/DiscretePID.h.
      p = struct('Kp', 1.0, 'Ki', 0.0, 'Kd', 0.0, 'N', 100.0, ...
                 'uMin', -1e9, 'uMax', 1e9, 'Kb', 1.0, 'b_weight', 1.0);
    end
  end
end
